ENH: special: add type stubs for `orthogonal.py`

person142 opened the pull request:
#### Reference issue
https://github.com/scipy/scipy/issues/11749

#### What does this implement/fix?
After adding the stubs, code like the following:

```python
import scipy.special as sc

reveal_type(sc.sh_jacobi(1, 1, 1))
reveal_type(sc.roots_legendre(1))
reveal_type(sc.roots_legendre(1, mu=True))
reveal_type(sc.roots_legendre(1, mu=False))
```

gives us the following types

```
scipy.special.orthogonal.orthopoly1d
Tuple[numpy.ndarray, numpy.ndarray]
Tuple[numpy.ndarray, numpy.ndarray, builtins.float]
Tuple[numpy.ndarray, numpy.ndarray]
```

Note that in particular we are correctly inferring when the roots
functions return a 2-tuple versus a 3-tuple.

The types were kept in stubs because getting the 2-versus-3 tuple
issue correct requires several overloads, so they are quite verbose.

#### Additional information
Note that this also adds `typing_extensions` to
`mypy_requirements.txt`. This is already included as a part of

- Mypy
- `numpy-stubs`

but we are now using it explicitly for `Literal`.

person142 commented on scipy/special/orthogonal.pyi:
```diff
@@ -0,0 +1,320 @@
+    'roots_sh_jacobi',
+]
+
+@overload
```
The verbose typing is needed to capture the fact that the `roots_*` functions return either 2 tuples or 3 tuples depending on the value of `mu`. The overloads are for

- default `mu`
- `mu = False`
- `mu = True`

You might think that we could get away with just the latter two cases, but we actually can't: if we try to do that then we have to do `mu: Literal[...] = ...`, and because of the `= ...` in the case where `mu` isn't specified Mypy has no way of determining what the right return type is.

rgommers commented on scipy/special/orthogonal.pyi:
```diff
@@ -0,0 +1,319 @@
+import numpy
+
+_IntegerType = Union[int, numpy.integer]
+_FloatingType = Union[float, numpy.floating]
```
It looks to me like these would also benefit from moving to `numpy-stubs`.

person142 replied:
Yeah, seems reasonable. Maybe the fact that we're already finding other types that would be useful to add is ammunition in favor of not putting types in the top-level NumPy namespace.

eric-wieser replied:
Could you use `numbers` instead here?
```
_IntegerType = numbers.Integral
_FloatingType = numbers.Real
```

person142 replied:
Ah yes, so generally

https://github.com/python/mypy/issues/3186

has made this not work very well.

rgommers commented:
I'm still missing something related to testing here. If I break something on purpose, `python runtests.py --mypy` is still happy, and:
```
$ mypy scipy
Success: no issues found in 630 source files
rgommers@machine:~/code/scipy (roots-types)$ mypy scipy/special
Success: no issues found in 71 source files
rgommers@machine:~/code/scipy (roots-types)$ mypy scipy/special/orthogonal.py
scipy/special/__init__.py:658: error: Cannot determine type of '__all__'  [has-type]
scipy/special/orthogonal.py:88: error: Module 'scipy.special' has no attribute 'specfun'  [attr-defined]
Found 2 errors in 2 files (checked 1 source file)
```
Whether or not one does `rm -rf .mypy_cache` also seems to matter sometimes.

Am I doing something wrong here, or is mypy known to be flaky like this?

person142 commented:
> If I break something on purpose, python runtests.py --mypy is still happy

Hm yes I'm getting the same; in particular I tried:

```
$ git diff scipy/special/tests/test_orthogonal.py
diff --git a/scipy/special/tests/test_orthogonal.py b/scipy/special/tests/test_orthogonal.py
index d4a7ee12a..d0aa82397 100644
--- a/scipy/special/tests/test_orthogonal.py
+++ b/scipy/special/tests/test_orthogonal.py
@@ -12,7 +12,7 @@ import scipy.special.orthogonal as orth

 class TestCheby(object):
     def test_chebyc(self):
-        C0 = orth.chebyc(0)
+        C0 = orth.chebyc('foo')
         C1 = orth.chebyc(1)
         olderr = np.seterr(all='ignore')
         try:
```

and am getting no complaints. A `reveal_type` in there tells us

```
scipy/special/tests/test_orthogonal.py:16: note: Revealed type is 'Any'
scipy/special/tests/test_orthogonal.py:16: note: 'reveal_type' always outputs 'Any' in unchecked functions
```

so the stubs aren't being picked up.

> Whether or not one does rm -rf .mypy_cache also seems to matter sometimes.

Yes, I've seen that too. Interestingly I've never seen that in other places where I use Mypy... I might have misconfigured something with our setup. One thing that might not be helping is that there's a cache in `build/path/to/build` but also one in the directory with the `mypy.ini`... maybe I should pin the cache directory to be one of those places.

person142 commented:
The funny thing is that for the types in the PR description I actually had the file similarly in `special/tests`, so sometimes the stubs are getting picked up there. I wonder if this is because the above tests do `import scipy.special.orthogonal as orth`?

person142 commented:
Ah ha! I needed to add an annotation to the tests to get mypy to check it fully:

```
$ git diff scipy/special/tests/test_orthogonal.py
diff --git a/scipy/special/tests/test_orthogonal.py b/scipy/special/tests/test_orthogonal.py
index d4a7ee12a..5302e0131 100644
--- a/scipy/special/tests/test_orthogonal.py
+++ b/scipy/special/tests/test_orthogonal.py
@@ -11,8 +11,8 @@ import scipy.special.orthogonal as orth


 class TestCheby(object):
-    def test_chebyc(self):
-        C0 = orth.chebyc(0)
+    def test_chebyc(self) -> None:
+        C0 = orth.chebyc('foo')
         C1 = orth.chebyc(1)
         olderr = np.seterr(all='ignore')
         try:
```

Now I get

```
scipy/special/tests/test_orthogonal.py:15: error: Argument 1 to "chebyc" has incompatible type "str"; expected "Union[int, integer]"  [arg-type]
```

person142 commented:
Oh derp, this was an obvious one; we don't have `check_untyped_defs` enabled, so any function without a type signature doesn't get checked. We probably don't want to enable that globally as it will likely give a cascade of errors, but we might want to start enabling it for stuff in `special` so that the tests exercise the type definitions too.

rgommers commented:
> We probably don't want to enable that globally as it will likely give a cascade of errors, but we might want to start enabling it for stuff in `special` so that the tests exercise the type definitions too.

That sounds good to me.

person142 commented:
Ok, I enabled `check_untyped_defs` for `test_orthogonal.py`, and lo and behold it caught a bunch of issues! Fixed those up; some notes:

- The tests followed the old pattern of using `orthogonal.eval_*` (this is kept around for back-compat reasons), but we really want people to be doing `sc.eval_*`. I didn't want to add stubs for the old-style `orthogonal.eval_*` style, so I updated the tests.
- Similar for `poch`-that function is exported from orthogonal just for back-compat reasons.

rgommers commented:
Nice, LGTM now. Thanks @person142

rth commented:
Thanks for working on this. @person142 It I wanted to try out typing that you are adding here and in other PRs e.g. from scikit-learn what would I need to do (besides of installing numpy-stubs and numpy/scipy master)?

person142 commented:
> what would I need to do

You can install `mypy_requirements.txt`:

https://github.com/scipy/scipy/blob/master/mypy_requirements.txt

and then SciPy from master. (Note that NumPy from master isn't required until such a time as we move the stubs into NumPy proper). The other thing is that we haven't marked any part of SciPy as PEP 561 compatible yet by adding the necessary `py.typed` file, so you would need to add that. If it helps with experimenting we could probably add it and then remove it if we don't make enough progress before the next release. Let me make an issue to discuss that...

person142 commented:
Issue about adding a `py.typed` file here: https://github.com/scipy/scipy/issues/11945.

rgommers commented:
> Yes, I've seen that too. Interestingly I've never seen that in other places where I use Mypy... I might have misconfigured something with our setup.

Turns out that it's a MyPy bug: https://github.com/python/mypy/issues/7777. There's ~60 open cache-related bugs, I don't think there's much we can do about it (except help out to fix them or `rm -rf .mypy_cache`'ing).
